python/http_ece/__init__.py

```diff
@@ -1,76 +1,138 @@
+    :type version: enumerate('aes128gcm', 'aesgcm', 'aesgcm128')
+
+    """
+    context = b""
+    keyinfo = ""
+    nonceinfo = ""
+
+    def build_info(base, info_context):
+        return b"Content-Encoding: " + base + b"\0" + info_context
+
+    def derive_dh(mode, keyid, dh, version="aesgcm"):
+
+        def length_prefix(key):
             return struct.pack("!H", len(key)) + key
 
         if keyid is None:
-            raise Exception(u"'keyid' is not specified with 'dh'")
-        if not keyid in keys:
-            raise Exception(u"'keyid' doesn't identify a key: " + keyid)
-        if not keyid in labels:
-            raise Exception(u"'keyid' doesn't identify a key label: " + keyid)
+            raise ECEException(u"'keyid' is not specified with 'dh'")
+        if keyid not in keys:
+            raise ECEException(u"'keyid' doesn't identify a key: " + keyid)
         if mode == "encrypt":
-            senderPubKey = keys[keyid].get_pubkey()
-            receiverPubKey = dh
+            sender_pub_key = key or keys[keyid].get_pubkey()
+            receiver_pub_key = dh
         elif mode == "decrypt":
-            senderPubKey = dh
-            receiverPubKey = keys[keyid].get_pubkey()
+            sender_pub_key = dh
+            receiver_pub_key = key or keys[keyid].get_pubkey()
         else:
-            raise Exception(u"unknown 'mode' specified: " + mode);
+            raise ECEException(u"unknown 'mode' specified: " + mode)
+        if version == "aes128gcm":
+            context = "WebPush: info\x00" + receiver_pub_key + sender_pub_key
+        else:
+            label = labels.get(keyid, 'P-256').encode('utf-8')
+            context = (label + b"\0" + length_prefix(receiver_pub_key) +
+                       length_prefix(sender_pub_key))
 
-        if type(labels[keyid]) == type(u""):
-            labels[keyid] = labels[keyid].encode("utf-8")
+        return keys[keyid].get_ecdh_key(dh), context
 
-        return (keys[keyid].get_ecdh_key(dh),
-                labels[keyid] + b"\0" +
-                    lengthPrefix(receiverPubKey) + lengthPrefix(senderPubKey))
+    # handle the older, ill formatted args.
+    pad_size = kwargs.get('padSize', 2)
+    auth_secret = kwargs.get('authSecret', auth_secret)
+    secret = key
 
-    if salt is None or len(salt) != 16:
-        raise Exception(u"'salt' must be a 16 octet value")
+    if version not in versions:
+        raise ECEException(u"invalid version specified")
 
-    context = b""
-    if key is not None:
-        secret = key
-    elif dh is not None:
-        (secret, context) = deriveDH(mode=mode, keyid=keyid, dh=dh)
-    elif keyid is not None:
-        secret = keys[keyid]
+    # handle old cases where version is explicitly None.
+    if not version:
+        if pad_size == 1:
+            version = "aesgcm128"
+        else:
+            version = "aesgcm"
+
+    if salt is None or len(salt) != 16:
+        raise ECEException(u"'salt' must be a 16 octet value")
```

Comment:
This is a bit complicated, sadly. Since we have to respect prior versions and function, I can't just gen up a new salt, (We used to throw errors, so we should still do so for older versions.)

Fortunately, in the `encrypt()` function we do check if salt has not been set, create one, and also force the content type to `aes128gcm`.